torch/csrc/jit/autodiff.cpp

```diff
@@ -164,6 +165,18 @@ bool isDifferentiable(Node* n) {
         static_cast<bool (*)(Node*)>(isDifferentiable));
   }
 
+  // check if the node inputs' value type is scalar type, in autodiff, scalar
+  // value aren't differentiable, so user should cast the scalar to double
+  // before using it.
```

Comment:
This looks wrong. `isDifferentiable` should never throw, but return `false` if it cannot prove that something can be differentiated for sure. What's the issue here anyway?

Reply:
for any node that calls `isDifferentiable`, we want to ensure their inputs are not Scalar/NumberType, because it make no sense for autodiff to differentiate a scalar(e.g. like if user call `t.item()` and pass it to an op, user should do explicit casting like `float()` to make it work in TorchScript and autodiff), so this is what we checks here. 

While we can just return `false` when we have this case, it would be better that we could warn the user the error message as well.

Reply:
How is it possible that we even compile such graphs? You can't use a `Scalar` as a `float` argument, because the scalar might be an int! So it's actually OK to return a `float` from backward, because `float` _is_ a `Scalar`, but forward is the problem.

Reply:
Yes, you can't use `Scalar` as a `float` argument. but shouldn't this `isDifferentiable` being called before we generate our `backward` graph (as I see it from [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/create_autodiff_subgraphs.cpp#L121))? the node that `isDifferentiable` checks is always the node in the forward graph. In the forward node, we need to avoid use a `Scalar` as a `Scalar` argument, user should always use `float` as a 'Scalar` argument. For example `tensor.clamp(t1.item(), t2.item())` this can be compiled but cannot be differentiated, user should write `tensor.clamp(float(t1.item()), float(t2.item()))` to make it differentiable in backward.

Reply:
I'm confused, how do we compile this case? Don't we just insert implicit casts? In that case it should be fine to return a float and the backward of a Scalar->float cast is a float->Scalar cast. From what I understand `isDifferentiable` is called on well-typed graphs, so you should never have to do such checks (we don't do that for other nodes FWIW).

Reply:
iirc we only insert implicit casts when the there's type mismatch and subtyping also happens, that of course will handle most of our cases, but there're some cases we don't do implicit casting and those scalars is something we need the user to do explicitly casting before use the variable. Taking a small example for `item()` call:

```python
@torch.jit.script
def test(a, min, max):
    return torch.clamp(a, min.item(), max.item())
```
this will turn into graph like:
```
graph(%a : Tensor,
      %min : Tensor,
      %max : Tensor):
  %3 : Scalar = aten::item(%min)
  %4 : Scalar = aten::item(%max)
  %5 : Tensor = aten::clamp(%a, %3, %4)
  return (%5)
```
we need to instruct the user to explicitly casting it before passing %3 or %4 to `clamp`, but as you said, we might do implicit casting for it or do the check in the compiler. I added the check here mainly because the above case will only become invalid if we want to differentiate the graph.

Reply:
Do we have a case where a Scalar is flowing to a Node, and we need the node to be symbolically differentiable for perf reasons? The example above with clamp probably doesn't happen in real code, where clamp is almost always called with constants. 

If the answer is no, then I recommend the check to be:
```
if (schema_type == NumberType::get() && input->type() != FloatType::get())
  return false; // formulas are only defined with floating point scalars, so we fallback to autograd for other cases.
```

As @apaszke noted, it is a bug to throw an exception here.

Reply:
we don't have a case right now, since function that have scalar as arg is rare. I am thinking of whether we should do the check in the `schema_matching` of compiler, I would suggest that we should never flow a `Scalar` to a node in TorchScript, rather we should instruct the user to do explicit type cast for scalar outputs. This might also fix the issue from bottom up, but it will make TorchScript syntax more strict.